### PR TITLE
Numeric subregion

### PR DESCRIPTION
**Pull request by jjb**
implements functionality discussed in #186

also an additional convenience function for displaying addresses, `Region#short_address_name`. let me know what you think of this method and the name!

### CODE REVIEW

File: lib/carmen/region.rb [hunk 1]
```diff
@@ -20,6 +20,18 @@ def name
       Carmen.i18n_backend.translate(path('name'))
     end
 
+    def numeric_code?
+      /[0-9]/ === code[0]
```
Comment (swcraig) [inline lib/carmen/region.rb]: Even though there aren't any cases currently where subregion codes mix digits and letters, I think I would prefer to see this cheek whether the entire string consists of digits (`/^[0-9]+$/ === code`).

If, for some crazy reason, there comes a time when subregion codes mix letters and digits, checking entirely for digits will make sure that we get consistent behaviour. (For example, what if there were codes 'M32' or '32M'?)

Comment (gr8bit) [reply lib/carmen/region.rb]: Agreed, may I suggest the bulletproof variant though: `/\A\d+\z/` :)

**Comment by jjb** [comment]
> For a country such as Bangladesh, the subregion codes are just single letters.  

interesting. something we could do is if a subregion code is a single letter, assume it's the serially-assigned variety like bangladesh, and in that case use the name. here's an example of a bagleshian address i found: https://www.unicef.org/infobycountry/bangladesh_bangladesh_contact.html

i'm guessing this will cover quite a lot of cases, and maybe all of them. BUT, maybe it's not nice

now, considering the above, and what you said about future codes possibly having letters, it makes me ponder the semantics of the whole thing anyway. i've been assuming that 2-letter codes for subregions are essentially always appropriate abbreviations. maybe this is actually much less common than i was assuming. so now i'm having an crisis about the whole PR :)

thoughts?

**Comment by jjb** [comment]
I can't seem to get the test data into the yaml. here's what i'm trying:


### locale/en/world/oc.yml

```yaml
---
en:
  world:
    oc:
      ao:
        name: Airstrip One
      '42':
        name: Numerica
      43N:
        name: Numerica Prime
      N44:
        name: Numerica Prime Prime
```

### data/world/oc.yml

```yaml
---
- code: AO
  type: province
- code: '42'
  type: province
- code: 43N
  type: province
- code: N44
  type: province
```

I also tried it with quotes around the codes, like with 42.

Without the alphanumeric mixed data, the tests were working (as seen in the PR).

**Comment by swcraig** [comment]
The way that Carmen accesses YAML data right now requires that the codes ("N44", "43N") in the locale directory be all lowercase, while their corresponding entries in the data directory be all uppercase. I believe that's why you are getting `nil` for the subregion in your previous comment.

**Comment by jjb** [comment]
that was it!

funny, i now realize that the discussion above suggested that 43n should /not/ be considered numeric. to me, that's a pretty ambiguous case. i don't know which i would prefer to print in an address. it matters what the convention is for the region.

i guess we can't get a perfect system until we start storing "abbreviation" in the data.

**Comment by jjb** [comment]
so i guess i would lean toward the idea that "43n" is probably not considered an abbreviation, so in that case i would want to use the full name in an address.

**Comment by jjb** [comment]
Okay, updated PR pushed, with 43n considered not numeric.

**Comment by swcraig** [comment]
I haven't been avoiding this PR by the way, I've been giving it some decent thought.

I like the intent of what this PR does, but I am starting to feel that semantically these methods aren't appropriate. I can't find a definitive source to say whether or not these two letter codes are suitable abbreviations (in terms of local customs), and even going under the assumption that they are, whether or not they are used in this way for addressing. `short_address_name` then becomes an ambiguous method where we have made an assumption as to what form is "useful".

Also related to this, in my (possibly contentious) opinion, Carmen should not concern itself _at all_ with the idea of "addressing". This library should focus on providing accurate and structured geographic data. If someone wants to use Carmen to write a library that provides correct address formatting for different regions/countries around the world, that's awesome, but it's not something that Carmen itself should provide.

**Comment by jjb** [comment]
I agree

so, `short_address_name` should be removed. i'm happy to also say that `numeric_code?` should also be the responsibility of the library user, since its purpose is probably for making decisions about presentation (like addresses).

feel free to close this

**Comment by swcraig** [comment]
I am glad that you feel the same way.  Leaving these presentation decisions to the user is best.

And thank you for the work on this!

**Pull request closed without merging**